Feature: Shopping List

rueblimaster opened the pull request:
## What
Adds a general shopping list, can be added to by other features etc. also can break down items into their recipes

<details>
<summary>Images</summary>
![image](https://github.com/user-attachments/assets/60e09f20-8a1f-4b8a-aee8-a410f3519886)
</details>

## Changelog New Features
+ Added Shopping List. - rueblimaster
## Changelog Technical Details
+ Added Renderables that work with click + modifiers. - rueblimaster

rueblimaster committed:
start with it

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
a little detekt

rueblimaster committed:
Merge branch 'shopping-list' of https://github.com/rueblimaster/SkyHanni into shopping-list

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
add ShoppingListItem.toString()

+ continued a little

rueblimaster committed:
start with the rly basic logic etc stuff

rueblimaster committed:
figure out rendering

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
move command back here + some detekt

rueblimaster committed:
refractor NEU to Neu

rueblimaster committed:
added most of the add/remove logic

rueblimaster committed:
reduce amount of returns in 1 function

rueblimaster committed:
continue

rueblimaster committed:
fix warnings

rueblimaster committed:
some detekt

rueblimaster committed:
even more detekt

rueblimaster committed:
make the for loop good

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
add recipe getting

at least partially

rueblimaster committed:
detekt

rueblimaster committed:
add showSummary Config option

to implement

rueblimaster committed:
make big progress with the recipe etc

also add more commands

rueblimaster committed:
detekt

rueblimaster committed:
stop shadowing of it

detekt

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
you can now select recipes and other functionality

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
began on stopping recursive recipes

but only kinda works like I want it to, I don't want you to be able to show dia blocks as recipe for diamonds

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
make it work!!

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
laid out my plans

rueblimaster committed:
start with the plans

rueblimaster committed:
detekt

rueblimaster committed:
make categories hidable

and other stuff

rueblimaster committed:
Update ShoppingListItem.kt

rueblimaster committed:
kinda did a summary of all items needed

rueblimaster committed:
summary kinda thing done

rueblimaster committed:
implement pinning and fixed hiding

also added tooltips to items

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
start with storage

return side of things still needs work...

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/config/storage/ProfileSpecificStorage.kt

hannibal002 committed:
merge conflict

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
idk why it's not working

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
still not working but idk why

rueblimaster committed:
Update ShoppingList.kt

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
fix it a little but still don't work?

rueblimaster committed:
redid storage, maybe now it works

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
made storage work!

rueblimaster committed:
reenable breakDownIntoSubitems

rueblimaster committed:
fixed not saving expanded recipes, replaced pinning

rueblimaster committed:
finish removing pinning

rueblimaster committed:
remove currentlyDecidingRecipe

rueblimaster committed:
implement left clicks

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
add Category option saveInStorage

rueblimaster committed:
start with own display item + other fixes

rueblimaster committed:
added possibly needed code

rueblimaster committed:
all the click actions work now

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
reworked clickable renderable stuff

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
continue with clicking

rueblimaster committed:
it's working!!!

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
detekt

rueblimaster committed:
Merge branch 'shopping-list' of https://github.com/rueblimaster/SkyHanni into shopping-list

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
final detekt

rueblimaster committed:
improve ClickTypeWithModifiers

rueblimaster committed:
classify TODOs

add now, soon or later and maybe or not maybe

rueblimaster committed:
detekt ShoppingListCategory.kt

rueblimaster committed:
detekt Renderable.kt

rueblimaster committed:
as always: some detekt

rueblimaster committed:
detekt ShoppingList.kt

rueblimaster committed:
reduce print statements and add user feedback messages

rueblimaster committed:
Update ShoppingListItem.kt

rueblimaster committed:
switch category to new renderable

rueblimaster committed:
add category.isEmpty()

rueblimaster committed:
add onDebugDataCollect()

rueblimaster committed:
reduce commented out code

rueblimaster committed:
small changes and better code

rueblimaster committed:
detekt

rueblimaster committed:
rewrite remove() for readability

rueblimaster committed:
item now uses internalName.createBuyTip()

rueblimaster committed:
detekt (once more...)

rueblimaster committed:
fixes

rueblimaster committed:
small fixes

rueblimaster committed:
implement more stuff

rueblimaster committed:
migrate visitor shopping list

rueblimaster committed:
remove unused code

rueblimaster committed:
fix stuff

also remove some commented out code

rueblimaster committed:
implement middle click

also split getRenderables()

rueblimaster committed:
reduce secondary constructors

rueblimaster committed:
detekt

rueblimaster committed:
moved all template conversion logic into templates

rueblimaster committed:
search? not working...

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
little oopsie

rueblimaster committed:
Merge branch 'shopping-list' of https://github.com/rueblimaster/SkyHanni into shopping-list

rueblimaster committed:
not much

rueblimaster committed:
improve test() contents

rueblimaster committed:
cache interalNames

rueblimaster committed:
small fixes in item

rueblimaster committed:
implement arrow up + down and other small improvements

rueblimaster committed:
make rendering stuff better

rueblimaster committed:
own detekt

rueblimaster committed:
detekt

rueblimaster committed:
massively improve commands

rueblimaster committed:
remove done TODO

rueblimaster committed:
create custom returned class CommandArguments

rueblimaster committed:
detekt

rueblimaster committed:
make CommandArguments private

rueblimaster committed:
detekt

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/features/garden/visitor/GardenVisitorFeatures.kt

rueblimaster committed:
remove CraftMaterialCollector

rueblimaster committed:
little improvement

rueblimaster committed:
make ItemsOverall more readable

rueblimaster committed:
make config option actually work...

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
fix merge

rueblimaster committed:
really fix merge (hopefully)

rueblimaster committed:
fix it?

rueblimaster committed:
I present you: the correct fix

and some minor code improvements

rueblimaster committed:
Merge pull request #3 from rueblimaster/fix-visitor-shopping-list-falsely-hiding

Fix: Visitor Shopping List disappearing while supercrafting

rueblimaster committed:
Revert "Fix: Visitor Shopping List disappearing while supercrafting"

rueblimaster committed:
Merge pull request #4 from rueblimaster/revert-3-fix-visitor-shopping-list-falsely-hiding

Revert "Fix: Visitor Shopping List disappearing while supercrafting"

rueblimaster committed:
finally: Fix: Visitor Shopping List disappearing while supercrafting

and other visitor shopping list fix

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/features/garden/visitor/GardenVisitorFeatures.kt

CalMWolfs committed:
fix merge

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/features/garden/visitor/GardenVisitorFeatures.kt

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
add config options

to implement

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/utils/InventoryUtils.kt

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/utils/InventoryUtils.kt

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
remove unneeded code change 1

rueblimaster committed:
revert integrations

rueblimaster committed:
remove unneeded TODO

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list-5

rueblimaster committed:
implement changes from main sl

rueblimaster committed:
revert ItemsOverall

rueblimaster committed:
revert storage

rueblimaster committed:
remove unused config toggles

rueblimaster committed:
detekt

rueblimaster committed:
remove stupid comments

rueblimaster committed:
compacting to reduce lines

rueblimaster committed:
remove useless test() command

rueblimaster committed:
detekt

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
fix import

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster commented:
@hannibal002 
What do I need to do, to get this pr looked at?

hannibal002 commented:
by me? wait until 3.0.0 is out.
by other devs, you can always ask if they want to look at it, but please dont be annoying and accept a no.

there is no point in splitting this up further, just dont create new feature prs over 300 lines from now on.
if you notice the thing you write gets bigger, stop writing and think what you wanna do. if its too many things, only do the first thing. im not talking about classes/apis, etc, im talking about user facing changes.

rueblimaster commented:
> by me? wait until 3.0.0 is out.
> 
> by other devs, you can always ask if they want to look at it, but please dont be annoying and accept a no.
> 
> 
> 
> there is no point in splitting this up further, just dont create new feature prs over 300 lines from now on.
> 
> if you notice the thing you write gets bigger, stop writing and think what you wanna do. if its too many things, only do the first thing. im not talking about classes/apis, etc, im talking about user facing changes.

By what I want to do, you mean functionality I want to add, right?

hannibal002 commented:
yes. splitting backend logic from the user facing feature doesnt help us testing it.

instead, splitting it in smaller, testable features would be ideal. 
once you write different things that might work good together and the pr gets too large stop. you can always add the infos what you wanna add in the future in the pr as well, but there is no point in writing the other feature yet as long as the first (and maybe even dependent) feature is not merged

rueblimaster commented:
> yes. splitting backend logic from the user facing feature doesnt help us testing it.
> 
> 
> 
> instead, splitting it in smaller, testable features would be ideal. 
> 
> once you write different things that might work good together and the pr gets too large stop. you can always add the infos what you wanna add in the future in the pr as well, but there is no point in writing the other feature yet as long as the first (and maybe even dependent) feature is not merged

Okay, I'll try
As you said changing this pr now doesn't make sense.
The problem I would've seen with this pr, is that it's not multiple features or so that work together, it's all part of 1 feature that needs all parts to work

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
updates

rueblimaster committed:
imports

rueblimaster committed:
detekt

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
Make position property immutable in ShoppingListConfig

Changed the 'position' property from 'var' to 'val' in ShoppingListConfig to make it immutable. This enhances safety by preventing unintended modifications to the position.

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
Merge branch 'beta' into shopping-list-3

rueblimaster committed:
Merge branch 'beta' into shopping-list-3
